### ISSUE TITLE
False negative of -Winteger-overflow

### ISSUE THREAD

**Issue by mmdriley** [llvm/llvm-project#46755]
|  |  |
| --- | --- |
| Bugzilla Link | [47411](https://llvm.org/bz47411) |
| Version | trunk |
| OS | All |
| CC | @zygoloid |

## Extended Description 
Repro at https://godbolt.org/z/1rorM4, code also copied below. See the assignment of `does_not_warn`.

Compiled with default options using clang version 12.0.0 (https://github.com/llvm/llvm-project.git 96ef6998dfcc4d7f2c7fcb87eb9369e94a40e127)

===

```cpp
#include <cstdint>
#include <iostream>
#include <utility>

struct HasAnInt64 {
    int64_t v;
};

int64_t g(HasAnInt64 s) { return 0; }

template<typename T>
int64_t g2(T s) { return 0; }

int main() {
    // Each of these raises a `-Winteger-overflow` diagnostic...
    int64_t warns1 = 512 * 1024 * 1024 * 1024;
    HasAnInt64 warns2{512 * 1024 * 1024 * 1024};
    int64_t warns3 = g({512 * 1024 * 1024 * 1024});
    int64_t warns4 = g2<HasAnInt64>({512 * 1024 * 1024 * 1024});

    // ... but not this one.
    auto does_not_warn = std::make_pair<int, HasAnInt64>(0, {512 * 1024 * 1024 * 1024});

    // Overflow occurs. This prints `0`.
    std::cout << does_not_warn.second.v << std::endl;

    return 0;
}
```
